File: backend/app/screen_perception/generated.py
from enum import Enum
from pydantic import BaseModel
from typing import List, Optional, Union
from uuid import UUID


class ChatDestination(Enum):
    CLOUD = "cloud"
    LOCAL = "local"


class MIMEType(Enum):
    IMAGE_JPEG = "image/jpeg"
    IMAGE_PNG = "image/png"


class Limits(BaseModel):
    allowed_mime_types: List[MIMEType]
    capture_timeout_ms: int
    max_bytes: int
    max_concurrency: int
    max_height: int
    max_pixels: int
    max_width: int
    request_timeout_ms: int
    snapshot_max_age_ms: int
    vision_timeout_ms: int


class ProtocolVersion(Enum):
    THE_10 = "1.0"


class RoutingDisclosureType(Enum):
    SCREEN_ROUTING_DISCLOSED = "screen_routing_disclosed"


class VisionDestination(Enum):
    CLOUD = "cloud"
    LOCAL = "local"
    UNCONFIGURED = "unconfigured"


class RoutingDisclosure(BaseModel):
    chat_destination: ChatDestination
    client_session_id: UUID
    event_id: UUID
    limits: Limits
    protocol_version: ProtocolVersion
    routing_revision: str
    type: RoutingDisclosureType
    vision_destination: VisionDestination


class ReasonCode(Enum):
    API_UNAVAILABLE = "api_unavailable"
    BACKEND_UNAVAILABLE = "backend_unavailable"
    BROWSER_SURFACE_REJECTED = "browser_surface_rejected"
    CAPTURE_NOT_ALLOWED = "capture_not_allowed"
    CAPTURE_OS_ERROR = "capture_os_error"
    CLOUD_CONSENT_REQUIRED = "cloud_consent_required"
    CONTEXT_MISMATCH = "context_mismatch"
    DUPLICATE_REQUEST = "duplicate_request"
    FRAME_UNAVAILABLE = "frame_unavailable"
    GENERATION_MISMATCH = "generation_mismatch"
    IMAGE_DECODE_FAILED = "image_decode_failed"
    IMAGE_TOO_LARGE = "image_too_large"
    IMAGE_TYPE_INVALID = "image_type_invalid"
    INSECURE_CONTEXT = "insecure_context"
    NO_CAPTURE_SOURCE = "no_capture_source"
    PICKER_CANCELLED = "picker_cancelled"
    REQUEST_CANCELLED = "request_cancelled"
    REQUEST_EXPIRED = "request_expired"
    REQUEST_NOT_FOUND = "request_not_found"
    ROUTING_CHANGED = "routing_changed"
    SESSION_EXPIRED = "session_expired"
    SESSION_NOT_FOUND = "session_not_found"
    SESSION_REVOKED = "session_revoked"
    SURFACE_MISMATCH = "surface_mismatch"
    SURFACE_UNKNOWN = "surface_unknown"
    TRANSIENT_ACTIVATION_REQUIRED = "transient_activation_required"
    VIDEO_TRACK_INVALID = "video_track_invalid"
    VISION_INVALID_RESPONSE = "vision_invalid_response"
    VISION_TIMEOUT = "vision_timeout"
    VISION_UNAVAILABLE = "vision_unavailable"
    VISION_UNCONFIGURED = "vision_unconfigured"
    VISION_UNSUPPORTED = "vision_unsupported"


class Stage(Enum):
    CAPTURE = "capture"
    CHAT = "chat"
    PICKER = "picker"
    SESSION = "session"
    SUPPORT = "support"
    UPLOAD = "upload"
    VISION = "vision"


class ScreenErrorType(Enum):
    SCREEN_ERROR = "screen_error"


class ScreenError(BaseModel):
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    reason_code: ReasonCode
    recoverable: bool
    stage: Stage
    type: ScreenErrorType
    request_id: Optional[UUID] = None
    screen_session_id: Optional[UUID] = None
    turn_id: Optional[UUID] = None


class CaptureState(Enum):
    ACTIVE = "active"
    OFF = "off"
    SELECTING = "selecting"
    UNAVAILABLE = "unavailable"
    UNSUPPORTED = "unsupported"


class RecognitionState(Enum):
    CAPTURING = "capturing"
    COMPOSING = "composing"
    FAILED = "failed"
    IDLE = "idle"
    RECOGNIZING = "recognizing"
    SNAPSHOT_REQUESTED = "snapshot_requested"
    SUCCEEDED = "succeeded"
    UPLOADING = "uploading"


class ScreenStatusType(Enum):
    SCREEN_STATUS = "screen_status"


class ScreenStatus(BaseModel):
    capture_state: CaptureState
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    recognition_state: RecognitionState
    type: ScreenStatusType
    last_recognized_capture_at: Optional[str] = None
    reason_code: Optional[ReasonCode] = None
    screen_session_id: Optional[UUID] = None


class SessionHeartbeatType(Enum):
    SCREEN_SESSION_HEARTBEAT = "screen_session_heartbeat"


class SessionHeartbeat(BaseModel):
    client_session_id: UUID
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    screen_session_id: UUID
    type: SessionHeartbeatType


class SessionHeartbeatAcceptedType(Enum):
    SCREEN_SESSION_HEARTBEAT_ACCEPTED = "screen_session_heartbeat_accepted"


class SessionHeartbeatAccepted(BaseModel):
    event_id: UUID
    generation: int
    lease_expires_at: str
    protocol_version: ProtocolVersion
    screen_session_id: UUID
    type: SessionHeartbeatAcceptedType


class SessionRevokeRequestedReason(Enum):
    BACKEND_DISCONNECT = "backend_disconnect"
    CAPTURE_ENDED = "capture_ended"
    CHARACTER_CHANGE = "character_change"
    CONSENT_REVOKED = "consent_revoked"
    CONVERSATION_CHANGE = "conversation_change"
    PAGEHIDE = "pagehide"
    TARGET_CHANGE = "target_change"
    USER_OFF = "user_off"


class SessionRevokeRequestedType(Enum):
    SCREEN_SESSION_REVOKE_REQUESTED = "screen_session_revoke_requested"


class SessionRevokeRequested(BaseModel):
    client_session_id: UUID
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    reason: SessionRevokeRequestedReason
    screen_session_id: UUID
    type: SessionRevokeRequestedType


class SessionRevokedReason(Enum):
    BACKEND_DISCONNECT = "backend_disconnect"
    BACKEND_RESTART = "backend_restart"
    CAPTURE_ENDED = "capture_ended"
    CHARACTER_CHANGE = "character_change"
    CONSENT_REVOKED = "consent_revoked"
    CONVERSATION_CHANGE = "conversation_change"
    LEASE_EXPIRED = "lease_expired"
    PAGEHIDE = "pagehide"
    ROUTING_CHANGED = "routing_changed"
    TARGET_CHANGE = "target_change"
    USER_OFF = "user_off"


class SessionRevokedType(Enum):
    SCREEN_SESSION_REVOKED = "screen_session_revoked"


class SessionRevoked(BaseModel):
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    reason: SessionRevokedReason
    screen_session_id: UUID
    type: SessionRevokedType


class ActualSurface(Enum):
    MONITOR = "monitor"
    WINDOW = "window"


class CloudConsent(BaseModel):
    cloud_derived_chat: bool
    cloud_vision: bool


class SessionStartRequestedType(Enum):
    SCREEN_SESSION_START_REQUESTED = "screen_session_start_requested"


class SessionStartRequested(BaseModel):
    actual_surface: ActualSurface
    character_id: str
    client_session_id: UUID
    cloud_consent: CloudConsent
    conversation_id: UUID
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    requested_surface: ActualSurface
    routing_revision: str
    type: SessionStartRequestedType


class SessionStartedType(Enum):
    SCREEN_SESSION_STARTED = "screen_session_started"


class SessionStarted(BaseModel):
    actual_surface: ActualSurface
    character_id: str
    client_session_id: UUID
    conversation_id: UUID
    event_id: UUID
    generation: int
    heartbeat_interval_ms: int
    lease_duration_ms: int
    lease_expires_at: str
    protocol_version: ProtocolVersion
    routing_revision: str
    screen_session_id: UUID
    type: SessionStartedType


class Source(Enum):
    EXPLICIT_UI = "explicit_ui"
    NATURAL_LANGUAGE_TEXT = "natural_language_text"
    NATURAL_LANGUAGE_VOICE = "natural_language_voice"


class SnapshotRequestedType(Enum):
    SCREEN_SNAPSHOT_REQUESTED = "screen_snapshot_requested"


class SnapshotRequested(BaseModel):
    capture_deadline: str
    event_id: UUID
    generation: int
    protocol_version: ProtocolVersion
    request_id: UUID
    requested_at: str
    screen_session_id: UUID
    source: Source
    turn_id: UUID
    type: SnapshotRequestedType


class SnapshotUploadAcceptedType(Enum):
    SCREEN_SNAPSHOT_UPLOAD_ACCEPTED = "screen_snapshot_upload_accepted"


class SnapshotUploadAccepted(BaseModel):
    event_id: UUID
    generation: int
    image_id: UUID
    protocol_version: ProtocolVersion
    received_at: str
    request_id: UUID
    screen_session_id: UUID
    type: SnapshotUploadAcceptedType


class SnapshotUploadMetadataType(Enum):
    SCREEN_SNAPSHOT_UPLOAD_METADATA = "screen_snapshot_upload_metadata"


class SnapshotUploadMetadata(BaseModel):
    actual_surface: ActualSurface
    byte_length: int
    captured_at: str
    client_session_id: UUID
    event_id: UUID
    generation: int
    height: int
    image_id: UUID
    mime_type: MIMEType
    protocol_version: ProtocolVersion
    request_id: UUID
    screen_session_id: UUID
    turn_id: UUID
    type: SnapshotUploadMetadataType
    width: int


ScreenPerceptionEvent = Union[
    RoutingDisclosure,
    SessionStartRequested,
    SessionStarted,
    SessionHeartbeat,
    SessionHeartbeatAccepted,
    SessionRevokeRequested,
    SessionRevoked,
    SnapshotRequested,
    SnapshotUploadMetadata,
    SnapshotUploadAccepted,
    ScreenStatus,
    ScreenError,
]

File: frontend/src/lib/screen-perception/generated.ts
/**
 * 画面共有の制御metadata契約。画像本文、質問本文、Vision観測本文、対象名は含めない。
 */
export type ScreenPerceptionEvent =
  | RoutingDisclosure
  | SessionStartRequested
  | SessionStarted
  | SessionHeartbeat
  | SessionHeartbeatAccepted
  | SessionRevokeRequested
  | SessionRevoked
  | SnapshotRequested
  | SnapshotUploadMetadata
  | SnapshotUploadAccepted
  | ScreenStatus
  | ScreenError;

export interface RoutingDisclosure {
    chat_destination:   ChatDestination;
    client_session_id:  string;
    event_id:           string;
    limits:             Limits;
    protocol_version:   "1.0";
    routing_revision:   string;
    type:               "screen_routing_disclosed";
    vision_destination: VisionDestination;
}

export type ChatDestination = "local" | "cloud";

export interface Limits {
    allowed_mime_types:  MIMEType[];
    capture_timeout_ms:  number;
    max_bytes:           number;
    max_concurrency:     number;
    max_height:          number;
    max_pixels:          number;
    max_width:           number;
    request_timeout_ms:  number;
    snapshot_max_age_ms: number;
    vision_timeout_ms:   number;
}

export type MIMEType = "image/png" | "image/jpeg";

export type VisionDestination = "local" | "cloud" | "unconfigured";

export interface ScreenError {
    event_id:           string;
    generation:         number;
    protocol_version:   "1.0";
    reason_code:        ReasonCode;
    recoverable:        boolean;
    request_id?:        string;
    screen_session_id?: string;
    stage:              Stage;
    turn_id?:           string;
    type:               "screen_error";
}

export type ReasonCode = "api_unavailable" | "insecure_context" | "transient_activation_required" | "capture_not_allowed" | "no_capture_source" | "capture_os_error" | "picker_cancelled" | "surface_mismatch" | "browser_surface_rejected" | "surface_unknown" | "video_track_invalid" | "frame_unavailable" | "image_type_invalid" | "image_too_large" | "image_decode_failed" | "session_not_found" | "session_revoked" | "session_expired" | "generation_mismatch" | "context_mismatch" | "request_not_found" | "request_expired" | "duplicate_request" | "cloud_consent_required" | "routing_changed" | "vision_unconfigured" | "vision_unsupported" | "vision_timeout" | "vision_unavailable" | "vision_invalid_response" | "request_cancelled" | "backend_unavailable";

export type Stage = "support" | "picker" | "capture" | "session" | "upload" | "vision" | "chat";

export interface ScreenStatus {
    capture_state:              CaptureState;
    event_id:                   string;
    generation:                 number;
    last_recognized_capture_at: null | string;
    protocol_version:           "1.0";
    reason_code:                ReasonCode | null;
    recognition_state:          RecognitionState;
    screen_session_id?:         null | string;
    type:                       "screen_status";
}

export type CaptureState = "off" | "selecting" | "active" | "unavailable" | "unsupported";

export type RecognitionState = "idle" | "snapshot_requested" | "capturing" | "uploading" | "recognizing" | "composing" | "succeeded" | "failed";

export interface SessionHeartbeat {
    client_session_id: string;
    event_id:          string;
    generation:        number;
    protocol_version:  "1.0";
    screen_session_id: string;
    type:              "screen_session_heartbeat";
}

export interface SessionHeartbeatAccepted {
    event_id:          string;
    generation:        number;
    lease_expires_at:  string;
    protocol_version:  "1.0";
    screen_session_id: string;
    type:              "screen_session_heartbeat_accepted";
}

export interface SessionRevokeRequested {
    client_session_id: string;
    event_id:          string;
    generation:        number;
    protocol_version:  "1.0";
    reason:            SessionRevokeRequestedReason;
    screen_session_id: string;
    type:              "screen_session_revoke_requested";
}

export type SessionRevokeRequestedReason = "user_off" | "target_change" | "conversation_change" | "character_change" | "capture_ended" | "pagehide" | "backend_disconnect" | "consent_revoked";

export interface SessionRevoked {
    event_id:          string;
    generation:        number;
    protocol_version:  "1.0";
    reason:            SessionRevokedReason;
    screen_session_id: string;
    type:              "screen_session_revoked";
}

export type SessionRevokedReason = "user_off" | "target_change" | "conversation_change" | "character_change" | "capture_ended" | "pagehide" | "backend_disconnect" | "consent_revoked" | "lease_expired" | "routing_changed" | "backend_restart";

export interface SessionStartRequested {
    actual_surface:    ActualSurface;
    character_id:      string;
    client_session_id: string;
    cloud_consent:     CloudConsent;
    conversation_id:   string;
    event_id:          string;
    generation:        number;
    protocol_version:  "1.0";
    requested_surface: ActualSurface;
    routing_revision:  string;
    type:              "screen_session_start_requested";
}

export type ActualSurface = "monitor" | "window";

export interface CloudConsent {
    cloud_derived_chat: boolean;
    cloud_vision:       boolean;
}

export interface SessionStarted {
    actual_surface:        ActualSurface;
    character_id:          string;
    client_session_id:     string;
    conversation_id:       string;
    event_id:              string;
    generation:            number;
    heartbeat_interval_ms: number;
    lease_duration_ms:     number;
    lease_expires_at:      string;
    protocol_version:      "1.0";
    routing_revision:      string;
    screen_session_id:     string;
    type:                  "screen_session_started";
}

export interface SnapshotRequested {
    capture_deadline:  string;
    event_id:          string;
    generation:        number;
    protocol_version:  "1.0";
    request_id:        string;
    requested_at:      string;
    screen_session_id: string;
    source:            Source;
    turn_id:           string;
    type:              "screen_snapshot_requested";
}

export type Source = "explicit_ui" | "natural_language_text" | "natural_language_voice";

export interface SnapshotUploadAccepted {
    event_id:          string;
    generation:        number;
    image_id:          string;
    protocol_version:  "1.0";
    received_at:       string;
    request_id:        string;
    screen_session_id: string;
    type:              "screen_snapshot_upload_accepted";
}

export interface SnapshotUploadMetadata {
    actual_surface:    ActualSurface;
    byte_length:       number;
    captured_at:       string;
    client_session_id: string;
    event_id:          string;
    generation:        number;
    height:            number;
    image_id:          string;
    mime_type:         MIMEType;
    protocol_version:  "1.0";
    request_id:        string;
    screen_session_id: string;
    turn_id:           string;
    type:              "screen_snapshot_upload_metadata";
    width:             number;
}
